src/goto-instrument/cover_basic_blocks.h

```diff
@@ -18,7 +18,46 @@ Author: Daniel Kroening
 
 class message_handlert;
 
-class cover_basic_blockst
+class cover_blocks_baset
+{
+public:
+  /// \param t a goto instruction
+  /// \return the block number of the block
+  ///         the given goto instruction is part of
+  virtual std::size_t block_of(goto_programt::const_targett t) const = 0;
+
+  /// \param block_nr a block number
+  /// \return  the instruction selected for
+  ///   instrumentation representative of the given block
+  virtual optionalt<goto_programt::const_targett>
+  instruction_of(std::size_t block_nr) const = 0;
+
+  /// \param block_nr a block number
+  /// \return  the source location selected for
+  ///   instrumentation representative of the given block
+  virtual const source_locationt &
+  source_location_of(std::size_t block_nr) const = 0;
+
+  /// Select an instruction to be instrumented for each basic block such that
+  /// the java bytecode indices for each basic block is unique
+  /// \param goto_program The goto program
+  /// \param message_handler The message handler
+  virtual void select_unique_java_bytecode_indices(
+    const goto_programt &goto_program,
+    message_handlert &message_handler){};
+
+  /// Outputs the list of blocks
+  virtual void output(std::ostream &out) const = 0;
+
+  /// Output warnings about ignored blocks
+  /// \param goto_program The goto program
+  /// \param message_handler The message handler
+  virtual void report_block_anomalies(
+    const goto_programt &goto_program,
+    message_handlert &message_handler){};
+};
+
+class cover_basic_blockst : public cover_blocks_baset
```

Comment:
You could make the language-independent instrumentation be done by the base class, i.e. there is then only a cover_basic_blockst and a derived cover_basic_blocks_javat.
Provide a virtual method for constructing the blocks instead of doing it in the constructor.